src/mem_table.rs

```diff
@@ -12,11 +12,73 @@ use crate::bytes_range::BytesRange;
 use crate::error::SlateDBError;
 use crate::iter::{KeyValueIterator, SeekToKey};
 use crate::merge_iterator::MergeIterator;
-use crate::types::{RowAttributes, RowEntry, ValueDeletable};
+use crate::types::RowEntry;
 use crate::utils::WatchableOnceCell;
 
+/// Memtable may contains multiple versions of a single user key, with a monotonically increasing sequence number.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub(crate) struct KVTableInternalKey {
+    user_key: Bytes,
+    seq: u64,
+}
+
+impl KVTableInternalKey {
+    pub fn new(user_key: Bytes, seq: u64) -> Self {
+        Self { user_key, seq }
+    }
+}
+
+impl Ord for KVTableInternalKey {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.user_key
+            .cmp(&other.user_key)
+            .then(self.seq.cmp(&other.seq).reverse())
+    }
+}
+
+impl PartialOrd for KVTableInternalKey {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub(crate) struct KVTableInternalKeyRange {
+    start_bound: Bound<KVTableInternalKey>,
+    end_bound: Bound<KVTableInternalKey>,
+}
+
+impl RangeBounds<KVTableInternalKey> for KVTableInternalKeyRange {
+    fn start_bound(&self) -> Bound<&KVTableInternalKey> {
+        self.start_bound.as_ref()
+    }
+
+    fn end_bound(&self) -> Bound<&KVTableInternalKey> {
+        self.end_bound.as_ref()
+    }
+}
+
+impl From<BytesRange> for KVTableInternalKeyRange {
+    fn from(range: BytesRange) -> Self {
+        let start_bound = match range.start_bound() {
+            Bound::Included(key) => Bound::Included(KVTableInternalKey::new(key.clone(), u64::MAX)),
```

Comment:
IIUC, if I had `KVTableInternalKeyRange::from(0, 99)`, the range would include `key=0, seq=u64_MAX` through `key=99, seq=0`. This seems counter intuitive to me. I would have expected `key=0, seq=0` until (excluding) `key=99, seq=0`.

1. Is my understanding correct?
2. If (1), why is the logic structured the way it is?

Reply:
the intention of this `impl From<BytesRange> for KVTableInternalKeyRange` is to make a  `KVTableInternalKeyRange` that includes the keys with all the possible seqnums in it. 

yes this logic around the `Exclude` boundary is a bit cumbersome, i can't recall the details at the moment, let me revisit this logic with some test cases...

Reply:
fixed the logic with added test cases in 37f1c1e, PTAL

Reply:
> the intention of this impl From<BytesRange> for KVTableInternalKeyRange is to make a KVTableInternalKeyRange that includes the keys with all the possible seqnums in it.

Gotcha. In that case, should it be:

```
let start_bound = match range.start_bound() {
    Bound::Included(key) => Bound::Included(KVTableInternalKey::new(key.clone(), **0**)),
    Bound::Excluded(key) => Bound::Excluded(KVTableInternalKey::new(key.clone(), **u64::MAX**)),
    Bound::Unbounded => Bound::Unbounded,
};
let end_bound = match range.end_bound() {
    Bound::Included(key) => Bound::Included(KVTableInternalKey::new(key.clone(), **u64::MAX**)),
    Bound::Excluded(key) => Bound::Excluded(KVTableInternalKey::new(key.clone(), **0**)),
    Bound::Unbounded => Bound::Unbounded,
};
```

If the start bound includes key, then it includes all rows with that key (regardless of their seqnum). If the lower bound _excludes_ the key, then it starts _after_ the last row with that key (regardless of their seqnum).

Similarly, if the end bound includes the key, then it includes all the rows with that key (regardless of their seqnum). And if the end bound excludes the key, then it includes all rows _before_ that key (regardless of their seqnum).

tl;dr I think the logic might need to be tweaked as I've shown in bold above?

Reply:
the range here is still a bit counter intuitive, given the seqnum is in a descremental order, `u64::MAX` is smaller than 0, and vice versa 😲.

given a key range of `(Included("key1"), Unbound)`, if we let the output as `(Included(InternalKey("key1", 0)), Unbound)`, `InternalKey("key1", 100)` would not be included in this range, which is not what we exepect.

Reply:
> given the seqnum is in a descremental order, u64::MAX is smaller than 0, and vice versa 😲.

Oh, right!! Man, this breaks my brain. 😵‍💫  Can you add some docs explaining this?